sui_programmability/examples/nfts/sources/collectible.move

```diff
@@ -0,0 +1,218 @@
+/// Custom metadata can be created and passed into the `Collectible` but that would
+/// require additional work on the creator side to set up metadata creation methods.
+module nfts::collectible {
+    use sui::object::{Self, UID};
+    use sui::publisher::{Self, Publisher};
+    use sui::tx_context::{TxContext};
+    use std::option::{Self, Option};
+    use std::string::String;
+    use std::vector as vec;
+
+    /// For when a witness type passed is not an OTW.
+    const ENotOneTimeWitness: u64 = 0;
+    /// For when the type `T` is not from the same module as the OTW.
+    const EModuleDoesNotContainT: u64 = 1;
+    /// For when maximum size of the Collection is reached - minting forbidden.
+    const ECapReached: u64 = 2;
+    /// For when Names length does not match `img_urls` length
+    const EWrongNamesLength: u64 = 3;
+    /// For when Descriptions length does not match `img_urls` length
+    const EWrongDescriptionsLength: u64 = 4;
+    /// For when Creators length does not match `img_urls` length
+    const EWrongCreatorsLength: u64 = 5;
+    /// For when Metadatas length does not match `img_urls` length
+    const EWrongMetadatasLength: u64 = 6;
+
+    /// Basic collectible - should contain only unique information (eg
+    /// if all collectibles have the same description, it should be put
+    /// into the Display to apply to all of the objects of this type, and
+    /// not in every object).
+    struct Collectible<T: store> has key, store {
+        id: UID,
+        /// The only required parameter for the Collectible.
+        /// Should only contain a unique part of the URL to be used in the
+        /// template engine in the `Display` and save gas and storage costs.
+        img_url: String,
+        name: Option<String>,
+        description: Option<String>,
+        creator: Option<String>,
+        meta: Option<T>
+    }
+
+    /// Capability granted to the collection creator. Is guaranteed to be one
+    /// per `T` in the `create_collection` function.
+    /// Contains the cap - maximum amount of Collectibles minted.
+    struct CollectionCreatorCap<phantom T: store> has key, store {
+        id: UID,
+        cap: Option<u64>,
+        minted: u64
+    }
+
+    /// Create a new collection and receive `CollectionCreatorCap` with a `Publisher`.
+    ///
+    /// To make sure that a collection is created only once, we require an OTW;
+    /// but since the collection also requires a Publisher to set up the display,
+    /// we create the Publisher object here as well.
+    ///
+    /// Type parameter `T` is phantom; so we constrain it via `Publisher` to be
+    /// defined in the same module as the OTW. Aborts otherwise.
+    public fun create_collection<OTW: drop, T: store>(
+        otw: OTW, cap: Option<u64>, ctx: &mut TxContext
+    ): (Publisher, CollectionCreatorCap<T>) {
+        assert!(sui::types::is_one_time_witness(&otw), ENotOneTimeWitness);
+
+        let pub = publisher::claim(otw, ctx);
```

Comment:
It seems a little strange to me that this function creates (and returns) the `Publisher` rather than accepting one by reference.

Reply:
(early return)

```suggestion
    public fun create_collection<T: store>(
        pub: &Publisher, cap: Option<u64>, ctx: &mut TxContext
    ): CollectionCreatorCap<T> {
        assert!(sui::types::is_one_time_witness(&otw), ENotOneTimeWitness);
```

Reply:
Then you can call it twice.

Reply:
We need to make sure CollectionCreatorCap gets only called once.

Reply:
Makes sense!  Thanks for clarifying :)